### PR TITLE
network relays management

### PR DESCRIPTION
**Pull request by codeswot**
## Description

Network Relays Management & Monitoring

  This PR implements comprehensive relay
  monitoring and management functionality,
  including a new relay monitor screen,
  enhanced tooltip system, and improved relay
  configuration options.

  🚀 Features Added

  Relay Monitor Screen

  - New dedicated monitoring interface
  (RelayMonitorScreen) accessible from the
  network settings
  - Real-time relay status display showing
  connection states (Connected, Connecting,
  Failed)
  - Pull-to-refresh functionality with custom
  loading indicators on both network and
  monitor screens
  - Comprehensive relay aggregation displaying
   normal, inbox, and key package relays in
  unified view

  Enhanced Tooltip System

  - Custom tooltip component (WnTooltip) with
  smart positioning and arrow indicators
  - Dynamic arrow positioning that adjusts
  horizontally to point toward trigger
  elements
  - Intelligent placement logic showing
  tooltips below by default, above when space
  is limited
  - Reusable footer support for flexible
  content like status legends
  - Auto-dismiss functionality with
  tap-to-close and 5-second timeout
  - Theme-aware styling with proper color
  handling for overlay contexts

  Relay Options Management

  - Relay options bottom sheet
  (RelayOptionsBottomSheet) for editing
  existing relays
  - Real-time URL validation with debounced
  input and error feedback
  - Clipboard paste functionality with smart
  protocol prefix handling
  - Relay function selection with checkbox
  interface for messaging, invites, and key
  exchange
  - Remove relay confirmation with
  user-friendly dialog

  Improved Code Organization

  - Extracted reusable components:
    - RelayTile widget for consistent relay
  display
    - WnRefreshingIndicator for loading states
    - WnStatusLegendItem for status indicators
  - Shared validation utilities
  (RelayValidation) for URL validation logic
  - Consistent widget lifecycle management
  with proper mounted checks

  🧪 Testing Coverage

  Comprehensive unit tests added for:
  - Relay URL validation logic with edge cases
   and protocol handling
  - Tooltip positioning algorithms including
  screen constraint handling
  - Refresh state management with loading
  states and error scenarios

  🔧 Technical Implementation

  Widget Lifecycle & Error Handling

  - Proper mounted checks preventing setState
  calls after widget disposal
  - Overlay management with safe removal and
  null checks for tooltips
  - Exception handling for async operations
  with user-friendly error messages

  Responsive Design

  - Smart layout adjustments for different
  screen sizes and orientations
  - Accessibility considerations with proper
  semantic structure

  State Management

  - Loading state coordination preventing
  concurrent refresh operations
  - Theme context handling with proper color
  extraction for overlays

  📋 Known Limitations

  ⚠️ API Integration Pending: The relay
  management operations (add, update, remove)
  currently show success messages but API 
  integration is not yet implemented. The
  following need backend integration:

  - Relay addition through the add relay
  bottom sheet
  - Relay updates via the options bottom sheet

  - Relay removal functionality
  - Actual relay connection status fetching
  - Real-time status updates

  🔄 Future Enhancements

  - Complete API integration for CRUD
  operations

  📁 Files Modified

  New Files:
  - lib/ui/settings/network/relay_monitor_scre
  en.dart
  - lib/ui/settings/network/relay_options_bott
  om_sheet.dart
  - lib/ui/settings/network/widgets/relay_tile
  .dart
  - lib/ui/core/ui/wn_tooltip.dart
  - lib/ui/core/ui/wn_status_legend_item.dart
  -
  lib/ui/core/ui/wn_refreshing_indicator.dart
  - lib/utils/relay_validation.dart
  - test/utils/relay_validation_test.dart
  - test/ui/core/ui/tooltip_positioning_test.d
  art
  - test/ui/settings/network/relay_refresh_log
  ic_test.dart

  Modified Files:
  -
  lib/ui/settings/network/network_screen.dart
  - Added monitor navigation and
  pull-to-refresh
  - lib/ui/settings/network/add_relay_bottom_s
  heet.dart - Refactored to use shared
  validation
  - lib/routing/routes.dart - Added monitor
  route constant
  - lib/routing/router_provider.dart - Added
  nested routing

  This implementation provides a solid
  foundation for relay management with
  comprehensive UI/UX patterns, proper error
  handling, and extensive test coverage, ready
   for API integration.

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

## Checklist

- [x] Run `just precommit` to ensure that formatting and linting are correct
- [ ] Updated the `CHANGELOG.md` file with your changes

### CODE REVIEW

**Comment by untreu2** [comment]
Can't pull-to-refresh be smoother? I mean, refreshing is possible, but when the user pulls, the animation should stay that way, and when they let go, it should go back.

https://github.com/user-attachments/assets/3b8c9198-2e8b-4f68-aa3c-5fb156981afd

**Comment by codeswot** [comment]
> I think the info dialog on this screen shouldn't prevent the navigation.
> image
> 
Nice catch

File: test/ui/core/ui/tooltip_positioning_test.dart [hunk 1]
```diff
@@ -0,0 +1,254 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+// Helper class to test tooltip positioning logic
+class TooltipPositioning {
```
Comment (josefinalliende) [inline test/ui/core/ui/tooltip_positioning_test.dart:5]: It looks like the only class being tested here is the one defined in this test file. I’m not sure I get the purpose of the test 🤔, it seems like the tooltip positioning could still be broken in the actual widget, and this test wouldn’t catch it… or am I missing something?

**Review by josefinalliende — COMMENTED** [review]
Huge work! 👏 Left some comments. I tried adding a relay but didn’t see anything happen, not sure if that’s expected since it’s still in progress or if it’s a bug.

A bit of feedback on the PR itself:
- Maybe a shorter summary with the feature’s context or the reason for the changes would be easier to follow
- This PR feels quite big to me, which makes reviewing harder.  Maybe splitting it into smaller chunks next time could make it much easier to review. 😉

**Comment by codeswot** [comment]
> Huge work! 👏 Left some comments. I tried adding a relay but didn’t see anything happen, not sure if that’s expected since it’s still in progress or if it’s a bug.
> 
> A bit of feedback on the PR itself:
> - Maybe a shorter summary with the feature’s context or the reason for the changes would be easier to follow
> - This PR feels quite big to me, which makes reviewing harder.  Maybe splitting it into smaller chunks next time could make it much easier to review. 😉 

The reason for the summary being long, it captures all the changes made and why. And for the reason it's not adding relays it's because the API is not completed yet and some flow will change for adding relays on the backend

**Comment by untreu2** [comment]
I get this error when I navigate back to chat list - **Relay Monitor > Network Settings > General Settings > !**
_Another exception was thrown: GoError: There is nothing
to pop_

I think you need to adjust the router provider.
https://github.com/user-attachments/assets/36048d94-0791-4042-b27a-b499027ad8b3

File: lib/ui/settings/network/widgets/network_section.dart [hunk 2]
```diff
@@ -215,4 +217,14 @@ class RelayInfo {
       status: status ?? this.status,
     );
   }
+
+  @override
+  bool operator ==(covariant RelayInfo other) {
```
[the remark this answers is not in the document]
Comment (josefinalliende) [reply lib/ui/settings/network/widgets/network_section.dart:224]: Oh, didn’t know that, thanks a lot for explaining! 🙌

File: test/ui/core/ui/tooltip_positioning_test.dart (hunk 1 above)
Comment (josefinalliende) [reply test/ui/core/ui/tooltip_positioning_test.dart:5]: But if this class isn’t actually used anywhere, I feel like the test might not give us much real protection. I’d probably lean toward not adding tests that don’t help us catch potential breakages… or am I missing something and this class is actually used in the widget?

Comment (codeswot) [reply test/ui/core/ui/tooltip_positioning_test.dart:5]: You are right, i missed that the positioning is calculated the same way as the class. i created a class for compact testing, i will remove the class or modify the actual widget to use the class methods instead of manually doing the calculations without the use of the class

**Review by josefinalliende — DISMISSED** [review]
✅  Great!!!
